### PR TITLE
feat(ONNXToTosa): support onnx.Expand through tosa.tile.

### PR DESCRIPTION
**Pull request by ttjost**
Spec for onnx.Expand: https://onnx.ai/onnx/operators/onnx__Expand.html

### CODE REVIEW

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 1]
```diff
@@ -962,3 +962,28 @@ func.func @test_less_broadcast(%arg0: tensor<13x21x1xf32>, %arg1: tensor<1xf32>)
...
+func.func @test_expand_splat(%arg0: tensor<1x64x1x1xf32>) -> tensor<64x64x64x64xf32> {
+  %0 = "onnx.Constant"() {value = dense<64> : tensor<4xi64>} : () -> tensor<4xi64>
+  %1 = "onnx.Expand"(%arg0, %0) {onnx_node_name = "Expand_335"} : (tensor<1x64x1x1xf32>, tensor<4xi64>) -> tensor<64x64x64x64xf32>
+  return %1 : tensor<64x64x64x64xf32>
+// CHECK-LABEL:  func @test_expand_splat
+// CHECK:           tosa.tile %[[VAL_0]] {multiples = array<i64: 64, 64, 64, 64>} : (tensor<1x64x1x1xf32>) -> tensor<64x64x64x64xf32>
+}
+
+func.func @test_expand_no_legalization(%arg0: tensor<1x64x1x1xf32>, %arg1: tensor<4xi64>) -> tensor<1x64x64x64xf32> {
+  %0 = "onnx.Expand"(%arg0, %arg1) {onnx_node_name = "Expand_335"} : (tensor<1x64x1x1xf32>, tensor<4xi64>) -> tensor<1x64x64x64xf32>
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:985]: Isn't the result type illegal, i.e. the ONNX op is illegal in itself? I think we should only have legal IR as input in tests.

File: test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir [hunk 2]
```diff
@@ -962,3 +962,28 @@ func.func @test_less_broadcast(%arg0: tensor<13x21x1xf32>, %arg1: tensor<1xf32>)
 // CHECK:           [[VAR_1_:%.+]] = tosa.greater [[VAR_0_]], %arg0 : (tensor<1x1x1xf32>, tensor<13x21x1xf32>) -> tensor<13x21x1xi1>
 // CHECK:           return [[VAR_1_]] : tensor<13x21x1xi1>
 }
+
+// -----
+
+func.func @test_expand(%arg0: tensor<1x64x1x1xf32>) -> tensor<1x64x64x64xf32> {
+  %0 = "onnx.Constant"() {value = dense<[1, 64, 64, 64]> : tensor<4xi64>} : () -> tensor<4xi64>
+  %1 = "onnx.Expand"(%arg0, %0) {onnx_node_name = "Expand_335"} : (tensor<1x64x1x1xf32>, tensor<4xi64>) -> tensor<1x64x64x64xf32>
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/Math/Elementwise.mlir:970]: Can you please remove the `{onnx_node_name = "Expand_335"}` here and below?

**Review by mgehre-amd — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp [hunk 3]
```diff
@@ -289,6 +290,32 @@ class ONNXReluOpLoweringToTOSA : public OpConversionPattern<ONNXReluOp> {
...
+    }
+
+    // Convert denseAttr to DenseI64ArrayAttr. This handles both splat and
+    // non-splat scenarios.
+    ArrayBuffer<WideNum> shapeWideNums = getElementsWideNums(denseAttr);
+    ArrayRef<int64_t> shapeArray =
+        castArrayRef<int64_t, WideNum>(shapeWideNums.get());
+    auto denseI64Attr = DenseI64ArrayAttr::get(op.getContext(), shapeArray);
+
+    rewriter.replaceOpWithNewOp<mlir::tosa::TileOp>(
```
Comment (cmcgirr-amd) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:314]: I don't think we can implement `onnx.Expand` generically with just the `tosa.TileOp`. AFAICS the tile op replicates the content of each dimension y amount of times. And onnx.Expand broadcast to that given shape. 

So the rank of `tosa.Tile` is the same for Input and Output, but onnx.Expand can have different rank at the output. And the shape array of onnx.Expand does not equal the `multiples` argument of the tileop. 

I think for unequal rank cases we need a reshape. So for:
```
Onnx.Expand(tensor<3x1>, [15, 3, 5]) -> tensor<15x3x5>
```
We would want:
```
x = tosa.reshape(tensor<3x1>, [1x3x1]) -> tensor<1x3x1>
tosa.reshape(y, [15, 1, 5]) -> tensor<15x3x5>
```

**Review by cmcgirr-amd — CHANGES_REQUESTED** [review]
I think we need to either restrict the conversion to the versions of expand that Tile can support directly or try to make it work for all scenarios.

**Comment by mgehre-amd** [comment]
@ttjost, is it likely that you will work on this PR soon?

**Comment by flemairen6** [comment]
Opened https://github.com/Xilinx/onnx-mlir/pull/150 to lower Expand using a different decomposition. I didn't reuse this PR because I basically restarted from zero.

**Pull request closed without merging**